sam/src/main/java/com/iluwatar/sam/StateEngine.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * The MIT License
+ * Copyright (c) 2014 Ilkka Seppälä
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package com.iluwatar.sam;
+
+/**
+ * Class defining state engine
+ *
+ */
+public class StateEngine {
+
+  private RocketState state;
+  private RocketView view;
+  private RocketActions actions;
+  
+  public StateEngine(RocketView view, RocketActions actions) {
+    this.view = view;
+    this.actions = actions;
+  }
+
+  /**
+   * Method for render model 
+   * 
+   * @param model from RocketModel
+   */
+  public void render(RocketModel model) {
+    if (model.getCounter() == 0) {

```

Comment:
Doubt: Half of our state i.e. Counter is inside Model and other half i.e. current state of rocket launcher is represented in terms of `RocketState` implementation. I think for current example it works, but how can we add abort action to this?
As soon as we introduce abort action, then we also introduce additional semantics that a RocketLauncher that is in Ready state / Launched state cannot be aborted. Also hypothetically if we get a start action after Abort we need to factor in the time required for technical team to check launchpad (say for 20 secs) and then again start the counter.